CMakeTests/CheckAndAddFlag.cmake

```diff
@@ -1,14 +1,50 @@
 include(CheckCCompilerFlag)
 include(CheckCXXCompilerFlag)
 
-macro(check_and_add_flag var flag)
-	check_c_compiler_flag(${flag} FLAG_C_${var})
-	if(FLAG_C_${var})
-		set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${flag}")
-	endif()
-
-	check_cxx_compiler_flag(${flag} FLAG_CXX_${var})
-	if(FLAG_CXX_${var})
-		set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${flag}")
-	endif()
-endmacro()
+# check_add_add_flag(<variable> <flag> [DEBUG | RELEASE])
+#
+# Add a C or C++ compilation flag to the current scope
+#
+# Can optionally add the flag to Debug or Release configurations only, use this when
+# targeting multi-configuration generators like Visual Studio or Xcode.
+# Release configurations means NOT Debug, so it will work for RelWithDebInfo or MinSizeRel too.
+#
+# If the flag is added successfully, the variables FLAG_C_<variable> and FLAG_CXX_<variable>
+# may be set to ON.
+#
+# Examples:
+#   check_and_add_flag(FOO -foo)
+#   check_and_add_flag(ONLYDEBUG -onlydebug DEBUG)
+#   check_and_add_flag(OPTMAX -O9001 RELEASE)
+
+function(check_and_add_flag var flag)
```

Comment:
I put a lot of comments on this, for clarity here's how it might look if you decided to go with all of them:

```cmake
function(check_and_add_flag var flag)
   set(genexp_config_test "1")
   if(ARGV2 STREQUAL "DEBUG_ONLY")
      set(genexp_config_test "$<CONFIG:Debug>")
   elseif(ARGV2 STREQUAL "RELEASE_ONLY")
      set(genexp_config_test "$<NOT:$<CONFIG:Debug>>")
   elseif(NOT ARGV2 STREQUAL "")
      message(FATAL_ERROR "check_and_add_flag called with incorrect arguments: ${ARGVN}")
   endif()

   check_c_compiler_flag(${flag} FLAG_C_${var})
   if(FLAG_C_${var})
      add_compile_options("$<$<AND:$<COMPILE_LANGUAGE:C>,${genexp_config_test}>:${flag}>")
   endif()

   check_cxx_compiler_flag(${flag} FLAG_CXX_${var})
   if(FLAG_CXX_${var})
      add_compile_options("$<$<AND:$<COMPILE_LANGUAGE:CXX>,${genexp_config_test}>:${flag}>")
   endif()
endfunction()
```